src/mmfatl.c

```diff
@@ -0,0 +1,1591 @@
+        }
+
+        result = addCheckOverflow(
+            result, strlen(aDescriptor->ellipsis) + 1 /* NUL */);
+    }
+    return result;
+}
+
+/*!
+ * returns a pointer to the first char of the memory where fatal error
+ * messages are kept, or built up.  The result is independent of any contents
+ * of the buffer.
+ * \return a pointer to the first character.
+ * \pre \ref descriptor is set up and \ref memBlock not NULL.
+ */
+static char* fatalErrorBufferBegin()
+{
+    return (char*)memBlock + descriptor.dmz;
+}
+
+/*!
+ * returns a pointer to the first char past the memory where fatal error
+ * messages are kept, or built up.  The result is independent of any contents
+ * of the buffer.
+ * \return a pointer to the first character behind the buffer, where the
+ *   ellipsis begins.
+ * \pre \ref descriptor is set up and \ref memBlock not NULL.
+ */
+static char* fatalErrorBufferEnd()
+{
+    return fatalErrorBufferBegin() + descriptor.size;
+}
+
+/*!
+ * reset the current contents of the fatal error buffer to the empty string.
+ * \pre \ref descriptor is set up and \ref memBlock not NULL.
+ * \post the contents of the buffer read as a NUL terminated string is
+ *   the empty string.
+ */
+static void clearFatalErrorBuffer()
+{
+    *fatalErrorBufferBegin() = NUL;
+}
+
+/*!
+ * initializes the buffer used for fatal error messages with a NUL terminated
+ * empty string.  Copies the ellipsis in \ref descriptor to the end of the
+ * buffer, so it is automatically appended in case of a message truncation.
+ * \pre \ref descriptor is set up and \ref memBlock not NULL.
+ * \post The ellipsis in \ref descriptor is copied to the \ref memBlock, and
+ *   the \ref descriptor is updated accordingly.  So no reference to a user
+ *   space ellipsis is kept in this code any more.
+ */
+static void initFatalErrorBuffer()
+{
+    clearFatalErrorBuffer();
+    /* This decouples from any user space memory, that can safely be changed or
+     * deallocated any time now */
+    descriptor.ellipsis = strcpy(fatalErrorBufferEnd(), descriptor.ellipsis);
+}
+
+void freeFatalErrorBuffer()
+{
+    FREE(memBlock);
+    memBlock = NULL;
+    descriptor.size = 0;
+    descriptor.dmz = 0;
+    descriptor.ellipsis = NULL;
+}
+
+int allocFatalErrorBuffer(struct FatalErrorBufferDescriptor const* aDescriptor)
+{
+    int result = FALSE;
+    freeFatalErrorBuffer();
+    size_t memSize = isValidFatalErrorBufferDescriptor(aDescriptor)?
+            evalRequestedMemSize(aDescriptor) : 0;
+    if (memSize > 0)
+    {
+        memBlock = MALLOC(memSize);
+        if (memBlock)
+        {
+            /* the assignment of the ellipsis is temporary only. */
+            descriptor = *aDescriptor;
+            initFatalErrorBuffer();
+            result = TRUE;
+        }
+    }
+    return  result;
+}
+
+char const* getFatalErrorMessage()
+{
+    return memBlock? fatalErrorBufferBegin() : NULL;
+}
+
+/*===================    Parsing the Format String   =====================*/
```

Comment:
And why are we implementing `printf` by hand now?

Reply:
Because I don't know what printf does, when you are in an out-of-memory situation, or even worse, your memory is corrupted.

Reply:
I don't think we should be reinventing aspects of the C standard library here. This is just more code to have to deal with for little gain. If this was actually fixing a present bug I would understand, but the harms here are entirely speculative.

Reply:
Hmm.  The goal is to guarantee Metamath verifies proofs as close to 100% as possible.  This covers reliable error detection and handling.  I could answer 'Assuming no harms is entirely speculative' as well.

Reply:
> Hmm. The goal is to guarantee Metamath verifies proofs as close to 100% as possible. This covers reliable error detection and handling.

While I agree in the abstract that we should be improving the reliability (and more importantly, maintainability) of metamath.exe, I'm not sure I agree with the extreme phrasing of "as close to 100% as possible". If we're talking metamath in general and not metamath.exe, then I believe the best way to get verification as close to 100% as possible (and here I do agree with the claim) is to not implement the verifier in C at all, and use some formal methods to construct the verifier and/or make use of really small verifiers that can easily be eyeballed for correctness.

But for metamath.exe, it is much more than a verifier, and it is also not a verifier on the critical path for the aforementioned reasons, so keeping the code under control is a much higher priority than correctness except in a few key places. That's not to say that correctness isn't important, but rather that it is one of multiple goals. It's not at the level where I would accept a bugfix for a theoretically predicted bug that makes the code more complicated and less clear.

> I could answer 'Assuming no harms is entirely speculative' as well.

We have a bug tracker if harms become more than speculative.

Reply:
> The use of format strings is entirely a convenience, and it's not so convenient if we have to maintain the parser too.

Exactly.  If you give up on the idea of formatted error messages in a tight or corrupt memory situation, you can dispose of half the code in this PR.  This is a sensible design decision.  But it breaks with the current implementation.

As for the parser: It has three states, so about 9 transitions (including operations not changing state) to watch.  This is really not out of the world.  The correctness is supervised by the regression tests.  I could dig out bison if this helps in such a small case.  My guess is: You have some pain checking the code in now, but I expect it to be stable for years afterwards.

And yes, radically stripping down complexity is an option, if not a necessity, judging by the resources the community is currently allocating.  Important questions.  Beyond my responsibility here.

Reply:
I don't know what to do here further.  The goals of this PR are written down under the title 'Rationale' in mmfatl.h, their implementation here is a standard conformant one.  The goals can be extended or reduced, but that is not this PR any longer then.